src/qdisc_funcs.sh

```diff
@@ -0,0 +1,636 @@
+                "egress") ret="$args31 rate 128kbit ceil ${prio_rate}kbit prio 1";;
+            esac;;
+        "3 2") ret="$args rate ${be_rate}kbit ceil ${be_ceil}kbit prio 2";;
+        "3 3") ret="$args rate ${bk_rate}kbit ceil ${be_ceil}kbit prio 3";;
+    esac
+
+    [ -n "$ret" ] && echo "$ret" && return
+
+    sqm_error "Unhandled args \"$dir $num_tiers $tier\" to $func in tier_shaper_$SHAPER()."
+    return 1
+}
+
+
+tier_shaper_hfsc() {
+
+    local func=$1
+    local dir=$2
+    local num_tiers=$3
+    local tier=$4
+    local rate
+
+    case $dir in
+        ingress)
+            rate="$DOWNLINK";;
+        egress)
+            rate="$UPLINK";;
+        *)
+            sqm_error "Unknown direction \"$dir\" in tier_shaper_$SHAPER()."
+            return 1;;
+    esac
+
+# Parameter comments below from Eric Luehrsen:
+
+# Link share (virtual time) tuning is only rough as borrowing will occur.
+# However in saturation, each class is guaranteed (real time) minimum of 15%.
+# This allows HFSC to do its work but avoids some of its vitrual time quirks.
+
+    local rate_c=$(( ${rate} * 50 / 49 ))
+    local rate_r=$(( ${rate} * 1 / 7 ))
+    local rate_1=$(( ${rate} * 2 / 7 ))
+    local rate_2=$(( ${rate} * 2 / 7 ))
+    local rate_3=$(( ${rate} * 3 / 7 ))
+
+    case $func in
+        flow_rate)
+            case "$num_tiers $tier" in
+                "1 1") echo "$rate";;
+                "3 1") echo "$rate_1";;
+                "3 2") echo "$rate_2";;
+                "3 3") echo "$rate_3";;
+                *)
+                    echo "$rate"
+                    sqm_warn "Unhandled args to flow_rate \"$num_tiers $tier\" in tier_shaper_$SHAPER(). Using defaults.";;
+            esac
+            return;;
+        default_tier)
+            case "$num_tiers" in
+                "1") echo "1";;
+                "3") echo "2";;
+                *)
+                    echo "$(expr $num_tiers - 1)"
+                    sqm_warn "Unhandled args to default_tier \"$num_tiers\" in tier_shaper_$SHAPER(). Using defaults.";;
+            esac
+            return;;
+        tier_args)
+            ;;
+        *)
+            sqm_error "Unknown function \"$func\" in tier_shaper_$SHAPER()."
+            return 1;;
+    esac
+
+# Root Class
+# The SC curve (LS+RT) includes feedback in virtual time, but the UL
+# ceiling is pure real time. If UL=SC, then you cant actually get SC.
+
+    local ret=""
+
+# CHECK/FIXME: Eric L.'s implementation has the leaf qdisc as child of the
+# root class "1 0)" but this framework requires the leaf to be a child of the
+# "1 1)" class. So need to choose some "safe" parameters to achieve this.
+
+    case "$num_tiers $tier" in
+        "1 root") ret="$(get_stab_string) $SHAPER";;
+        "1 0") ret="$SHAPER sc m1 ${rate_c}kbit d 1s m2 ${rate}kbit ul rate ${rate_c}kbit";;
+        "1 1") ret="$SHAPER ls rate ${rate}kbit ul rate ${rate_c}kbit";;
```

Comment:
Confirmed Structure: HFSC doesn't require subclasses like others but should allow the structure to be the same. In the root SC parameter, "rate" is the closed loop average target. UL is effectively integrator wind up protection to meet that target. SC starts at "rate_c" and tapers over 1 [s] back to "rate" to prime the integrator. Otherwise, it tends to tip-toe up to speed.

Fix Parameter: the leaf doesn't require UL (or RT) and these real-time limits may cause grief with only one subclass. LS is the initial guess for sharing virtual-time bandwidth. It can be the lone parameter, and should be the same "rate," because it has no one to share with.